cluster_utils/settings.py

```diff
@@ -314,74 +383,165 @@ def add_cmd_line_params(base_dict, extra_flags):
-    :param verbose: Boolean to determine if final settings are pretty printed
-    :return: Settings object with (deep) dot access.
+def read_main_script_params_with_smart_settings(
+    settings_file: pathlib.Path,
+    cmdline_settings: Optional[list[str]] = None,
+    make_immutable: bool = True,
+    dynamic: bool = True,
+    pre_unpack_hooks: Optional[list] = None,
+    post_unpack_hooks: Optional[list] = None,
+) -> smart_settings.AttributeDict:
+    """Read parameters for the cluster_utils main script using smart_settings.
+
+    Args:
+        settings_file:  Path to the settings file.
+        cmdline_settings:  List of additional parameters provided via command line.
+        make_immutable:  See ``smart_settings.load()``
+        dynamic:  See ``smart_settings.load()``
+        pre_unpack_hooks:  See ``smart_settings.load()``
+        post_unpack_hooks:  See ``smart_settings.load()``
+
+    Returns:
+        Parameters as loaded by smart_settings.
     """
+    cmdline_settings = cmdline_settings or []
     pre_unpack_hooks = pre_unpack_hooks or []
     post_unpack_hooks = post_unpack_hooks or []
 
+    if not is_settings_file(os.fspath(settings_file)):
+        raise ValueError(f"{settings_file} is not a supported settings file.")
+
+    def add_cmd_params(orig_dict):
+        add_cmd_line_params(orig_dict, cmdline_settings)
+
+    return smart_settings.load(
+        os.fspath(settings_file),
+        make_immutable=make_immutable,
+        dynamic=dynamic,
+        post_unpack_hooks=([add_cmd_params, check_reserved_params] + post_unpack_hooks),
+        pre_unpack_hooks=pre_unpack_hooks,
+    )
+
+
+def init_job_script_argument_parser() -> argparse.ArgumentParser:
+    """Initialise ArgumentParser for job scripts."""
+
+    def server_info(ip_and_port: str) -> dict[str, str | int]:
+        """Split and validate string in "ip:port" format.  For use with argparse."""
+        ip, port = ip_and_port.rsplit(":", maxsplit=1)
+        if not port.isdigit():
+            raise ValueError("Invalid port")
+        return {"ip": ip, "port": int(port)}
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "parameter_file_or_dict",
+        type=str,
+        help="""
+        Path to a configuration file or (if `--dict` is set) a string defining a Python
+        dictionary with the parameters.
+    """,
+    )
+    parser.add_argument(
+        "parameter_overwrites",
+        nargs="*",
+        type=str,
+        default=[],
+        metavar="<key>=<value>",
+        help="""Additional parameters in the format '<key>=<value>'.  Values provided
+            here overwrite parameters provided via `parameter_file_or_dict`.  Key has to
+            match a configuration option, value has to be a valid Python literal.
+            Example: `'results_dir="/tmp"' 'optimization_setting.run_local=True'`
+        """,
+    )
+    parser.add_argument(
+        "--dict",
```

Comment:
Can we name this as `--parameter-dict`? Just `--dict` is a little non-descriptive.